File: csharp/AppEncryption/AppEncryption.Tests/AppEncryption/Persistence/DynamoDbMetastoreImplTest.cs
using System;
using System.Diagnostics.CodeAnalysis;
using System.Threading;
using System.Threading.Tasks;
using Amazon.DynamoDBv2;
using Amazon.DynamoDBv2.DocumentModel;
using Amazon.Runtime;
using GoDaddy.Asherah.AppEncryption.Persistence;
using GoDaddy.Asherah.AppEncryption.Tests.Fixtures;
using GoDaddy.Asherah.Crypto.Exceptions;
using Microsoft.Extensions.Logging;
using Moq;
using Newtonsoft.Json.Linq;
using Xunit;

namespace GoDaddy.Asherah.AppEncryption.Tests.AppEncryption.Persistence
{
    [ExcludeFromCodeCoverage]
    public class DynamoDbMetastoreImplTest : IClassFixture<DynamoDbContainerFixture>, IClassFixture<MetricsFixture>, IDisposable
    {
        private const string PartitionKey = "Id";
        private const string SortKey = "Created";
        private const string AttributeKeyRecord = "KeyRecord";

        private const string Region = "us-west-2";

        private readonly AmazonDynamoDBClient _amazonDynamoDbClient;

        private readonly DynamoDbMetastoreImpl _dynamoDbMetastoreImpl;
        private readonly DateTimeOffset _created;
        private string _serviceUrl;

        public DynamoDbMetastoreImplTest(DynamoDbContainerFixture dynamoDbContainerFixture)
        {
            _serviceUrl = dynamoDbContainerFixture.GetServiceUrl();
            var clientConfig = new AmazonDynamoDBConfig
            {
                ServiceURL = _serviceUrl,
                AuthenticationRegion = "us-west-2",
            };
            _amazonDynamoDbClient = new AmazonDynamoDBClient(clientConfig);

            DynamoDbMetastoreHelper.CreateTableSchema(_amazonDynamoDbClient, "EncryptionKey").Wait();

            _dynamoDbMetastoreImpl = DynamoDbMetastoreImpl.NewBuilder(Region)
                .WithEndPointConfiguration(_serviceUrl, Region)
                .Build();

            // Pre-populate test data using helper and capture the created timestamp
            _created = DynamoDbMetastoreHelper.PrePopulateTestDataUsingOldMetastore(_amazonDynamoDbClient, "EncryptionKey", Region).Result;
        }

        public void Dispose()
        {
            try
            {
                _ = _amazonDynamoDbClient.DeleteTableAsync(_dynamoDbMetastoreImpl.TableName).Result;
            }
            catch (AggregateException)
            {
                // There is no such table.
            }
        }


        [Fact]
        public void TestLoadSuccess()
        {
            var actualJsonObject = _dynamoDbMetastoreImpl.Load(DynamoDbMetastoreHelper.ExistingTestKey, _created);

            Assert.True(actualJsonObject.IsSome);
            Assert.True(JToken.DeepEquals(JObject.FromObject(DynamoDbMetastoreHelper.ExistingKeyRecord), (JObject)actualJsonObject));
        }

        [Fact]
        public void TestLoadWithNoResultShouldReturnEmpty()
        {
            var actualJsonObject = _dynamoDbMetastoreImpl.Load("fake_key", _created);

            Assert.False(actualJsonObject.IsSome);
        }

        [Fact]
        public void TestLoadWithFailureShouldReturnEmpty()
        {
            Dispose();
            var actualJsonObject = _dynamoDbMetastoreImpl.Load(DynamoDbMetastoreHelper.ExistingTestKey, _created);

            Assert.False(actualJsonObject.IsSome);
        }

        [Fact]
        public void TestLoadLatestWithSingleRecord()
        {
            var actualJsonObject = _dynamoDbMetastoreImpl.LoadLatest(DynamoDbMetastoreHelper.ExistingTestKey);

            Assert.True(actualJsonObject.IsSome);
            Assert.True(JToken.DeepEquals(JObject.FromObject(DynamoDbMetastoreHelper.ExistingKeyRecord), (JObject)actualJsonObject));
        }

        [Fact]
        public void TestLoadLatestWithSingleRecordAndSuffix()
        {
            var dbMetastoreImpl = DynamoDbMetastoreImpl.NewBuilder(Region)
                .WithEndPointConfiguration(_serviceUrl, Region)
                .WithKeySuffix()
                .Build();

            var actualJsonObject = dbMetastoreImpl.LoadLatest(DynamoDbMetastoreHelper.ExistingTestKey);

            Assert.True(actualJsonObject.IsSome);
            Assert.True(JToken.DeepEquals(JObject.FromObject(DynamoDbMetastoreHelper.ExistingKeyRecord), (JObject)actualJsonObject));
        }

        [Fact]
        public async Task TestLoadLatestWithMultipleRecords()
        {
            // Create a local table instance for this test
            var table = new TableBuilder(_amazonDynamoDbClient, _dynamoDbMetastoreImpl.TableName)
                .AddHashKey(PartitionKey, DynamoDBEntryType.String)
                .AddRangeKey(SortKey, DynamoDBEntryType.Numeric)
                .Build();

            var createdMinusOneHour = _created.AddHours(-1);
            var createdPlusOneHour = _created.AddHours(1);
            var createdMinusOneDay = _created.AddDays(-1);
            var createdPlusOneDay = _created.AddDays(1);

            // intentionally mixing up insertion order
            var documentPlusOneHour = new Document
            {
                [PartitionKey] = DynamoDbMetastoreHelper.ExistingTestKey,
                [SortKey] = createdPlusOneHour.ToUnixTimeSeconds(),
                [AttributeKeyRecord] = Document.FromJson(new JObject
                {
                    { "mytime", createdPlusOneHour },
                }.ToString()),
            };
            await table.PutItemAsync(documentPlusOneHour, CancellationToken.None);

            var documentPlusOneDay = new Document
            {
                [PartitionKey] = DynamoDbMetastoreHelper.ExistingTestKey,
                [SortKey] = createdPlusOneDay.ToUnixTimeSeconds(),
                [AttributeKeyRecord] = Document.FromJson(new JObject
                {
                    { "mytime", createdPlusOneDay },
                }.ToString()),
            };
            await table.PutItemAsync(documentPlusOneDay, CancellationToken.None);

            var documentMinusOneHour = new Document
            {
                [PartitionKey] = DynamoDbMetastoreHelper.ExistingTestKey,
                [SortKey] = createdMinusOneHour.ToUnixTimeSeconds(),
                [AttributeKeyRecord] = Document.FromJson(new JObject
                {
                    { "mytime", createdMinusOneHour },
                }.ToString()),
            };
            await table.PutItemAsync(documentMinusOneHour, CancellationToken.None);

            var documentMinusOneDay = new Document
            {
                [PartitionKey] = DynamoDbMetastoreHelper.ExistingTestKey,
                [SortKey] = createdMinusOneDay.ToUnixTimeSeconds(),
                [AttributeKeyRecord] = Document.FromJson(new JObject
                {
                    { "mytime", createdMinusOneDay },
                }.ToString()),
            };
            await table.PutItemAsync(documentMinusOneDay, CancellationToken.None);

            var actualJsonObject = _dynamoDbMetastoreImpl.LoadLatest(DynamoDbMetastoreHelper.ExistingTestKey);

            Assert.True(actualJsonObject.IsSome);
            Assert.True(JToken.DeepEquals(createdPlusOneDay, ((JObject)actualJsonObject).GetValue("mytime")));
        }

        [Fact]
        public void TestLoadLatestWithNoResultShouldReturnEmpty()
        {
            var actualJsonObject = _dynamoDbMetastoreImpl.LoadLatest("fake_key");

            Assert.False(actualJsonObject.IsSome);
        }

        [Fact]
        public void TestLoadLatestWithFailureShouldReturnEmpty()
        {
            Dispose();
            var actualJsonObject = _dynamoDbMetastoreImpl.LoadLatest(DynamoDbMetastoreHelper.ExistingTestKey);

            Assert.False(actualJsonObject.IsSome);
        }

        [Fact]
        public void TestStore()
        {
            var actualValue = _dynamoDbMetastoreImpl.Store(DynamoDbMetastoreHelper.ExistingTestKey, DateTimeOffset.Now, JObject.FromObject(DynamoDbMetastoreHelper.ExistingKeyRecord));

            Assert.True(actualValue);
        }

        [Fact]
        public void TestStoreWithSuffixSuccess()
        {
            var dbMetastoreImpl = DynamoDbMetastoreImpl.NewBuilder(Region)
                .WithEndPointConfiguration(_serviceUrl, Region)
                .WithKeySuffix()
                .Build();
            var actualValue = dbMetastoreImpl.Store(DynamoDbMetastoreHelper.ExistingTestKey, DateTimeOffset.Now, JObject.FromObject(DynamoDbMetastoreHelper.ExistingKeyRecord));

            Assert.True(actualValue);
        }

        [Fact]
        public void TestStoreWithClientProvidedExternally()
        {
            var client = new AmazonDynamoDBClient(new AmazonDynamoDBConfig
            {
                ServiceURL = _serviceUrl,
                AuthenticationRegion = Region,
            });

            var dbMetastoreImpl = DynamoDbMetastoreImpl.NewBuilder(Region)
                .WithDynamoDbClient(client)
                .Build();
            var actualValue = dbMetastoreImpl.Store(DynamoDbMetastoreHelper.ExistingTestKey, DateTimeOffset.Now, JObject.FromObject(DynamoDbMetastoreHelper.ExistingKeyRecord));

            Assert.True(actualValue);
        }

        [Fact]
        public void TestStoreWithDbErrorShouldThrowException()
        {
            Dispose();
            Assert.Throws<AppEncryptionException>(() =>
                _dynamoDbMetastoreImpl.Store(DynamoDbMetastoreHelper.ExistingTestKey, DateTimeOffset.Now, JObject.FromObject(DynamoDbMetastoreHelper.ExistingKeyRecord)));
        }

        [Fact]
        public void TestStoreWithDuplicateShouldReturnFalse()
        {
            var now = DateTimeOffset.Now;
            var firstAttempt = _dynamoDbMetastoreImpl.Store(DynamoDbMetastoreHelper.ExistingTestKey, now, JObject.FromObject(DynamoDbMetastoreHelper.ExistingKeyRecord));
            var secondAttempt = _dynamoDbMetastoreImpl.Store(DynamoDbMetastoreHelper.ExistingTestKey, now, JObject.FromObject(DynamoDbMetastoreHelper.ExistingKeyRecord));

            Assert.True(firstAttempt);
            Assert.False(secondAttempt);
        }

        [Fact]
        public void TestBuilderPathWithEndPointConfiguration()
        {
            var dbMetastoreImpl = DynamoDbMetastoreImpl.NewBuilder(Region)
                .WithEndPointConfiguration(_serviceUrl, Region)
                .Build();

            Assert.NotNull(dbMetastoreImpl);
        }

        [Fact]
        public void TestBuilderPathWithRegion()
        {
            var builder = new Mock<DynamoDbMetastoreImpl.Builder>(Region);
            var loadTable = (Table)new TableBuilder(_amazonDynamoDbClient, "EncryptionKey")
                .AddHashKey(PartitionKey, DynamoDBEntryType.String)
                .AddRangeKey(SortKey, DynamoDBEntryType.Numeric)
                .Build();

            builder.Setup(x => x.LoadTable(It.IsAny<IAmazonDynamoDB>(), Region))
                .Returns(loadTable);

            var dbMetastoreImpl = builder.Object
                .WithRegion(Region)
                .Build();

            Assert.NotNull(dbMetastoreImpl);
        }

        [Fact]
        public void TestBuilderPathWithKeySuffix()
        {
            var dbMetastoreImpl = DynamoDbMetastoreImpl.NewBuilder(Region)
                .WithEndPointConfiguration(_serviceUrl, Region)
                .WithKeySuffix()
                .Build();

            Assert.NotNull(dbMetastoreImpl);
            Assert.Equal(Region, dbMetastoreImpl.GetKeySuffix());
        }

        [Fact]
        public void TestBuilderPathWithoutKeySuffix()
        {
            var dbMetastoreImpl = DynamoDbMetastoreImpl.NewBuilder(Region)
                .WithEndPointConfiguration(_serviceUrl, Region)
                .Build();

            Assert.NotNull(dbMetastoreImpl);
            Assert.Equal(string.Empty, dbMetastoreImpl.GetKeySuffix());
        }

        [Fact]
        public void TestBuilderPathWithCredentials()
        {
            var dbMetastoreImpl = DynamoDbMetastoreImpl.NewBuilder(Region)
                .WithEndPointConfiguration(_serviceUrl, Region)
                .WithCredentials(new BasicAWSCredentials("dummykey", "dummy_secret"))
                .Build();

            Assert.NotNull(dbMetastoreImpl);
        }

        [Fact]
        public void TestBuilderPathWithInvalidCredentials()
        {
            var emptySecretKey = string.Empty;
            Assert.ThrowsAny<Exception>(() => DynamoDbMetastoreImpl.NewBuilder(Region)
                .WithEndPointConfiguration(_serviceUrl, Region)
                .WithCredentials(new BasicAWSCredentials("not-dummykey", emptySecretKey))
                .Build());
        }

        [Fact]
        public async Task TestBuilderPathWithTableName()
        {
            const string tempTableName = "DummyTable";

            // Use AWS SDK to create client
            var amazonDynamoDbConfig = new AmazonDynamoDBConfig
            {
                ServiceURL = _serviceUrl,
                AuthenticationRegion = "us-west-2",
            };
            var tempDynamoDbClient = new AmazonDynamoDBClient(amazonDynamoDbConfig);
            await DynamoDbMetastoreHelper.CreateTableSchema(tempDynamoDbClient, tempTableName);

            // Put the object in temp table
            var tempTable = (Table)new TableBuilder(tempDynamoDbClient, tempTableName)
                .AddHashKey(PartitionKey, DynamoDBEntryType.String)
                .AddRangeKey(SortKey, DynamoDBEntryType.Numeric)
                .Build();
            var jObject = JObject.FromObject(DynamoDbMetastoreHelper.ExistingKeyRecord);
            var document = new Document
            {
                [PartitionKey] = DynamoDbMetastoreHelper.ExistingTestKey,
                [SortKey] = _created.ToUnixTimeSeconds(),
                [AttributeKeyRecord] = Document.FromJson(jObject.ToString()),
            };
            await tempTable.PutItemAsync(document, CancellationToken.None);

            // Create a metastore object using the withTableName step
            var dbMetastoreImpl = DynamoDbMetastoreImpl.NewBuilder(Region)
                .WithEndPointConfiguration(_serviceUrl, "us-west-2")
                .WithTableName(tempTableName)
                .Build();
            var actualJsonObject = dbMetastoreImpl.Load(DynamoDbMetastoreHelper.ExistingTestKey, _created);

            // Verify that we were able to load and successfully decrypt the item from the metastore object created withTableName
            Assert.True(actualJsonObject.IsSome);
            Assert.True(JToken.DeepEquals(JObject.FromObject(DynamoDbMetastoreHelper.ExistingKeyRecord), (JObject)actualJsonObject));
        }

        [Fact]
        public void TestPrimaryBuilderPath()
        {
            var builder = new Mock<DynamoDbMetastoreImpl.Builder>(Region);
            var loadTable = (Table)new TableBuilder(_amazonDynamoDbClient, "EncryptionKey")
                .AddHashKey(PartitionKey, DynamoDBEntryType.String)
                .AddRangeKey(SortKey, DynamoDBEntryType.Numeric)
                .Build();

            builder.Setup(x => x.LoadTable(It.IsAny<IAmazonDynamoDB>(), Region))
                .Returns(loadTable);

            var dbMetastoreImpl = builder.Object
                .Build();

            Assert.NotNull(dbMetastoreImpl);
        }

        [Fact]
        public void TestBuilderPathWithLoggerEnabled()
        {
            var mockLogger = new Mock<ILogger>();

            var dbMetastoreImpl = DynamoDbMetastoreImpl.NewBuilder(Region)
                .WithEndPointConfiguration(_serviceUrl, Region)
                .WithLogger(mockLogger.Object)
                .Build();

            Assert.NotNull(dbMetastoreImpl);
        }

        [Fact]
        public void TestBuilderPathWithLoggerDisabled()
        {
            var dbMetastoreImpl = DynamoDbMetastoreImpl.NewBuilder(Region)
                .WithEndPointConfiguration(_serviceUrl, Region)
                .Build();

            Assert.NotNull(dbMetastoreImpl);
        }

        [Fact]
        public void TestBuilderPathWithLoggerAndCredentials()
        {
            var mockLogger = new Mock<ILogger>();

            var dbMetastoreImpl = DynamoDbMetastoreImpl.NewBuilder(Region)
                .WithEndPointConfiguration(_serviceUrl, Region)
                .WithLogger(mockLogger.Object)
                .WithCredentials(new BasicAWSCredentials("dummykey", "dummy_secret"))
                .Build();

            Assert.NotNull(dbMetastoreImpl);
        }

        [Fact]
        public void TestWithLoggerReturnsCorrectInterface()
        {
            var mockLogger = new Mock<ILogger>();

            var buildStep = DynamoDbMetastoreImpl.NewBuilder(Region)
                .WithEndPointConfiguration(_serviceUrl, Region)
                .WithLogger(mockLogger.Object);

            Assert.IsAssignableFrom<DynamoDbMetastoreImpl.IBuildStep>(buildStep);
        }
    }
}

File: java/app-encryption/src/test/java/com/godaddy/asherah/appencryption/persistence/DynamoDbMetastoreImplTest.java
package com.godaddy.asherah.appencryption.persistence;

import java.math.BigDecimal;
import java.net.URI;
import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.Map;
import java.util.Optional;

import com.amazonaws.auth.AWSStaticCredentialsProvider;
import com.amazonaws.auth.BasicAWSCredentials;
import com.amazonaws.client.builder.AwsClientBuilder;
import com.amazonaws.services.dynamodbv2.AmazonDynamoDB;
import com.amazonaws.services.dynamodbv2.AmazonDynamoDBClientBuilder;
import com.amazonaws.services.dynamodbv2.document.DynamoDB;
import com.amazonaws.services.dynamodbv2.document.Item;
import com.amazonaws.services.dynamodbv2.document.Table;
import org.json.JSONObject;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import com.amazonaws.services.dynamodbv2.local.main.ServerRunner;
import com.amazonaws.services.dynamodbv2.local.server.DynamoDBProxyServer;
import com.godaddy.asherah.appencryption.exceptions.AppEncryptionException;
import com.google.common.collect.ImmutableMap;
import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
import software.amazon.awssdk.auth.credentials.AwsCredentialsProvider;
import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.services.dynamodb.DynamoDbClient;
import software.amazon.awssdk.services.dynamodb.model.*;

import static com.godaddy.asherah.appencryption.persistence.DynamoDbMetastoreImpl.*;
import static org.junit.jupiter.api.Assertions.*;

class DynamoDbMetastoreImplTest {

  static final String DYNAMO_DB_PORT = "8000";
  static final String TEST_KEY = "some_key";
  static final String REGION = "us-west-2";
  static final String TEST_KEY_WITH_SUFFIX = "some_key_" + REGION;
  static DynamoDBProxyServer server;

  final Map<String, ?> keyRecord = ImmutableMap.of(
      "ParentKeyMeta", ImmutableMap.of(
          "KeyId", "_SK_api_ecomm",
          "Created", 1541461380L),
      "Key", "mWT/x4RvIFVFE2BEYV1IB9FMM8sWN1sK6YN5bS2UyGR+9RSZVTvp/bcQ6PycW6kxYEqrpA+aV4u04jOr",
      "Created", 1541461381L);

  final Instant instant = Instant.now().minus(1, ChronoUnit.DAYS);

  final AwsCredentialsProvider testCredentialsProvider = StaticCredentialsProvider.create(
      AwsBasicCredentials.create("test", "test"));

  DynamoDbMetastoreImpl dynamoDbMetastoreImpl;
  DynamoDbClient dynamoDbClient;
  String tableName;

  @BeforeAll
  public static void setupClass() throws Exception {
    System.setProperty("sqlite4java.library.path", "target/native-libs");
    String port = DYNAMO_DB_PORT;
    server = ServerRunner.createServerFromCommandLineArgs(
        new String[]{"-inMemory", "-port", port});
    server.start();
  }

  @AfterAll
  public static void teardownClass() throws Exception {
    server.stop();
  }

  @BeforeEach
  void setUp() {
    dynamoDbClient = DynamoDbClient.builder()
      .region(Region.of(REGION))
      .endpointOverride(URI.create("http://localhost:" + DYNAMO_DB_PORT))
      .credentialsProvider(testCredentialsProvider)
      .build();
    dynamoDbMetastoreImpl = DynamoDbMetastoreImpl.newBuilder(REGION, dynamoDbClient)
      .build();
    tableName = dynamoDbMetastoreImpl.getTableName();

    // Create table schema
    createTableSchema(dynamoDbClient, dynamoDbMetastoreImpl.getTableName());

    putItem(TEST_KEY, instant.getEpochSecond(), keyRecord);
    putItem(TEST_KEY_WITH_SUFFIX, instant.getEpochSecond(), keyRecord);
  }

  public void createTableSchema(final DynamoDbClient dynamoDbClient, final String tableName) {
    // Create table schema
    dynamoDbClient.createTable(request ->
      request
        .tableName(tableName)
        .keySchema(
          KeySchemaElement.builder()
            .attributeName(PARTITION_KEY)
            .keyType(KeyType.HASH)
            .build(),
          KeySchemaElement.builder()
            .attributeName(SORT_KEY)
            .keyType(KeyType.RANGE)
            .build())
        .attributeDefinitions(
          AttributeDefinition.builder()
            .attributeName(PARTITION_KEY)
            .attributeType(ScalarAttributeType.S)
            .build(),
          AttributeDefinition.builder()
            .attributeName(SORT_KEY)
            .attributeType(ScalarAttributeType.N)
            .build())
        .provisionedThroughput(ProvisionedThroughput.builder()
          .readCapacityUnits(1L)
          .writeCapacityUnits(1L)
          .build()));
  }

  @AfterEach
  void tearDown() {
    // Blow out the whole table so we have clean slate each time
    dynamoDbClient.deleteTable(request -> request.tableName(dynamoDbMetastoreImpl.getTableName()));
  }

  @Test
  void testLoadSuccess() {
    Optional<JSONObject> actualJsonObject = dynamoDbMetastoreImpl.load(TEST_KEY, instant);

    assertTrue(actualJsonObject.isPresent());
    assertEquals(keyRecord, actualJsonObject.get().toMap());
  }

  @Test
  void testLoadWithNoResultShouldReturnEmpty() {
    Optional<JSONObject> actualJsonObject = dynamoDbMetastoreImpl.load("fake_key", Instant.now());

    assertFalse(actualJsonObject.isPresent());
  }
  @Test
  void testLoadWithFailureShouldReturnEmpty() {
    Optional<JSONObject> actualJsonObject = dynamoDbMetastoreImpl.load(null, Instant.now());

    assertFalse(actualJsonObject.isPresent());
  }

  @Test
  void testLoadLatestWithSingleRecord() {
    Optional<JSONObject> actualJsonObject = dynamoDbMetastoreImpl.loadLatest(TEST_KEY);

    assertTrue(actualJsonObject.isPresent());
    assertEquals(keyRecord, actualJsonObject.get().toMap());
  }

  @Test
  void testLoadLatestWithSingleRecordAndSuffix() {
    DynamoDbMetastoreImpl dynamoDbMetastoreImpl = DynamoDbMetastoreImpl.newBuilder(REGION, dynamoDbClient)
      .withKeySuffix()
      .build();
    Optional<JSONObject> actualJsonObject = dynamoDbMetastoreImpl.loadLatest(TEST_KEY);

    assertTrue(actualJsonObject.isPresent());
    assertEquals(keyRecord, actualJsonObject.get().toMap());
  }

  @Test
  void testLoadLatestWithMultipleRecords() {
    Instant instantMinusOneHour = instant.minus(1, ChronoUnit.HOURS);
    Instant instantPlusOneHour = instant.plus(1, ChronoUnit.HOURS);
    Instant instantMinusOneDay = instant.minus(1, ChronoUnit.DAYS);
    Instant instantPlusOneDay = instant.plus(1, ChronoUnit.DAYS);

    // intentionally mixing up insertion order
    putItem(TEST_KEY, instantPlusOneHour.getEpochSecond(), ImmutableMap.of(
      "mytime", instantPlusOneHour.getEpochSecond()));
    putItem(TEST_KEY, instantMinusOneDay.getEpochSecond(), ImmutableMap.of(
      "mytime", instantMinusOneDay.getEpochSecond()));
    putItem(TEST_KEY, instantMinusOneHour.getEpochSecond(), ImmutableMap.of(
      "mytime", instantMinusOneHour.getEpochSecond()));
    putItem(TEST_KEY, instantPlusOneDay.getEpochSecond(), ImmutableMap.of(
      "mytime", instantPlusOneDay.getEpochSecond()));

    Optional<JSONObject> actualJsonObject = dynamoDbMetastoreImpl.loadLatest(TEST_KEY);

    assertTrue(actualJsonObject.isPresent());
    assertEquals(instantPlusOneDay.getEpochSecond(), actualJsonObject.get().getLong("mytime"));
  }

  @Test
  void testLoadLatestWithNoResultShouldReturnEmpty() {
    Optional<JSONObject> actualJsonObject = dynamoDbMetastoreImpl.loadLatest("fake_key");

    assertFalse(actualJsonObject.isPresent());
  }

  @Test
  void testLoadLatestWithFailureShouldReturnEmpty() {
    Optional<JSONObject> actualJsonObject = dynamoDbMetastoreImpl.loadLatest(null);

    assertFalse(actualJsonObject.isPresent());
  }

  @Test
  void testStoreSuccess() {
    DynamoDbMetastoreImpl dynamoDbMetastoreImpl = DynamoDbMetastoreImpl.newBuilder(REGION, dynamoDbClient)
      .withKeySuffix()
      .build();
    boolean actualValue = dynamoDbMetastoreImpl.store(TEST_KEY, Instant.now(), new JSONObject(keyRecord));

    assertTrue(actualValue);
  }

  @Test
  void testStoreWithSuffixSuccess() {
    boolean actualValue = dynamoDbMetastoreImpl.store(TEST_KEY, Instant.now(), new JSONObject(keyRecord));

    assertTrue(actualValue);
  }

  @Test
  void testStoreWithDuplicateShouldReturnFalse() {
    Instant now = Instant.now();
    boolean firstAttempt = dynamoDbMetastoreImpl.store(TEST_KEY, now, new JSONObject(keyRecord));
    boolean secondAttempt = dynamoDbMetastoreImpl.store(TEST_KEY, now, new JSONObject(keyRecord));

    assertTrue(firstAttempt);
    assertFalse(secondAttempt);
  }

  @Test
  void testStoreWithFailureShouldThrowException() {
    assertThrows(AppEncryptionException.class,
        () -> dynamoDbMetastoreImpl.store(null, Instant.now(), new JSONObject()));
  }

  @Test
  void testBuilderPathWithKeySuffix() {
    DynamoDbMetastoreImpl dynamoDbMetastore = DynamoDbMetastoreImpl.newBuilder(REGION, dynamoDbClient)
      .withKeySuffix()
      .build();

    assertEquals(REGION, dynamoDbMetastore.getKeySuffix());
  }

  @Test
  void testBuilderPathWithoutKeySuffix() {
    DynamoDbMetastoreImpl dynamoDbMetastore = DynamoDbMetastoreImpl.newBuilder(REGION, dynamoDbClient)
      .build();

    assertEquals("", dynamoDbMetastore.getKeySuffix());
  }

  @Test
  void testBuilderPathWithTableName() {
    String tableName = "DummyTable";

    createTableSchema(dynamoDbClient, tableName);

    // Put the object in DummyTable
    putItem(tableName, TEST_KEY, instant.getEpochSecond(), keyRecord);

    // Create a metastore object using the withTableName step
    DynamoDbMetastoreImpl dynamoDbMetastore = DynamoDbMetastoreImpl.newBuilder(REGION, dynamoDbClient)
      .withTableName(tableName)
      .build();
    Optional<JSONObject> actualJsonObject = dynamoDbMetastore.load(TEST_KEY, instant);

    // Verify that we were able to load and successfully decrypt the item from the
    //metastore object created withTableName
    assertTrue(actualJsonObject.isPresent());
    assertEquals(keyRecord, actualJsonObject.get().toMap());
  }

  @Test
  void testBackwardCompatibility() {
    // use sdk v1 to store keyrecord
    AmazonDynamoDB client = AmazonDynamoDBClientBuilder.standard()
      .withCredentials(new AWSStaticCredentialsProvider(
        new BasicAWSCredentials("test", "test")))
      .withEndpointConfiguration(
        new AwsClientBuilder.EndpointConfiguration("http://localhost:8000", "us-west-2"))
      .build();

    Table table = new DynamoDB(client).getTable(tableName);

    final Map<String, ?> keyRecordBeforeUpgrade = ImmutableMap.of(
      "ParentKeyMeta", ImmutableMap.of(
        "KeyId", "_SK_api_ecomm",
        "Created", new BigDecimal(1541461380L)),
      "Key", "mWT/x4RvIFVFE2BEYV1IB9FMM8sWN1sK6YN5bS2UyGR+9RSZVTvp/bcQ6PycW6kxYEqrpA+aV4u04jOr",
      "Created", new BigDecimal(1541461380L));

    Item item = new Item()
      .withPrimaryKey(
        PARTITION_KEY, "backward-test-key",
        SORT_KEY, instant.getEpochSecond())
      .withMap(ATTRIBUTE_KEY_RECORD, keyRecordBeforeUpgrade);
    table.putItem(item);

    // use sdk v2 to load keyrecord
    JSONObject actualJsonObject = dynamoDbMetastoreImpl.load("backward-test-key", instant).orElse(null);

    // verify
    assertNotNull(actualJsonObject);
    assertEquals(actualJsonObject.get("Key"), keyRecordBeforeUpgrade.get("Key"));
    assertEquals(actualJsonObject.get("Created"), ((BigDecimal) keyRecordBeforeUpgrade.get("Created")).longValueExact());
  }


  private void putItem(String partitionKey, Long sortKey, Map<String, ?> keyRecord) {
    putItem(tableName, partitionKey, sortKey, keyRecord);
  }

  private void putItem(String tableName, String partitionKey, Long sortKey, Map<String, ?> keyRecord) {
    dynamoDbClient.putItem(request ->
      request
        .tableName(tableName)
        .item(Map.of(
          PARTITION_KEY, AttributeValue.fromS(partitionKey),
          SORT_KEY, AttributeValue.fromN(Long.toString(sortKey)),
          ATTRIBUTE_KEY_RECORD, AttributeValue.fromM(DynamoDbMetastoreImpl.toDynamoDbItem(new JSONObject(keyRecord)))
        ))
    );
  }
}
